collectors/UcsmCollector.py

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/python3
+# -*- coding: utf-8 -*-
+
+import traceback
+import json
+from prometheus_client.core import GaugeMetricFamily
+from modules.ucsm import UcsmServer
+from collectors.ServerLicenseCollector import UcsServerLicenseCollector
+from BaseCollector import BaseCollector
+from ucsmsdk.ucsexception import UcsException
+
+
+class UcsmCollector(BaseCollector):
+    def __init__(self, creds, inventory_file):
+        self.creds = creds
+        self.inventory_file = inventory_file
+        self.handles = dict()
+
+    def describe(self):
+        yield GaugeMetricFamily("ucsm_metrics", "ucsm_collector_registered")
+
+    def get_inventory(self):
+        """
+        Get updated inventory each time collection happens
+        :return:
+        """
+        with open(self.inventory_file) as json_file:
+            data = json.load(json_file)
+        return data["servers"]
+
+    def collect(self):
+        print("Get updated inventory !")
+        server_list = self.get_inventory()
+
+        g = GaugeMetricFamily('ucsm_exporter_info', 'information metric, internal exporter info')
+
+        print("Create handles for Ucsm Servers.")
+        for x in range(len(server_list)):
```

Comment:
The handle handling part should actually go to your BaseClass. There you do the login. Always call a BaseClass variable which is holding the handle. Could easily be your dict. Therefore you would be able to use it more than once. 
If the handle is not working anymore, you need to reinitiate the login procedure. For now the easiest way is to implement a retry function in case the handle is not working properly. (here: `LicenceCollectore  mo = handle.query_dn(dn)`

Catch the error which occurs with an expired handle and then have a the login function called again to overwrite the old handle.

A better implementation however would be, to provide some sort of an interface. We take care of this after you've implemented this.

Reply:
@viennaa what I am thinking is
1. Handle handling in BaseClass: The handle for ucsm and ucsc differes like, ucsm needs UcsHandle handle from ucsmsdk, and ucsc needs UcscHandle from ucscsdk. then for login, handle.login doesnt accept the timeout for ucsc. So considering this I was thinking to have these handles in UcsmCollector and UcscCollector. let me know, what you think?

If you think to have it in Baseclass then we need to handle that in login function.

2. Regarding handle reinitiation, If you check in L41-L48 I am checking for the handle if exists, then again I am doing login on the same. I checked this works even if the VPN breaks and then we can go ahead and collect the metrics. I will surely handle the exception at query_dn. Let me know if think this is what you want.

Reply:
1. I am just seeing `UcsHandle` being used: https://github.com/sapcc/ucs-exporter/pull/1/files#diff-6d0b49c065163976a8157419af4bbc03R18
no sign of the UcsmHandle you are referring to.
I just designed a bigger draft, how this could work with some more classes, but I get the feeling it is maybe a little bit too much, so we go with the handle approach like you have it.

2. not really, but it is hard to implement with the current structure I guess. But now that I am looking at it again, you are triggering a login at each collect() run again (L42), since the condition always matches after it has been initialised initially. Ok guess that should work without a retry. 

3. what about closing the connection at the end, shouldn't there be some logout? I guess the handle timeout is something like 1h or so, so it might lock up further loggins, if we have to many (lose) open handles.